rules-java/api/src/main/java/org/jboss/windup/rules/apps/mavenize/ModuleAnalysisHelper.java

```diff
@@ -0,0 +1,149 @@
+import org.jboss.windup.graph.model.ProjectModel;
+import org.jboss.windup.graph.model.WindupVertexFrame;
+import org.jboss.windup.rules.apps.java.config.ScanPackagesOption;
+import org.jboss.windup.rules.apps.java.model.JavaClassFileModel;
+import org.jboss.windup.rules.apps.java.model.project.MavenProjectModel;
+import org.jboss.windup.util.Logging;
+
+/**
+ *
+ * @author <a href="http://ondra.zizka.cz/">Ondrej Zizka, ozizka at seznam.cz</a>
+ */
+public class ModuleAnalysisHelper
+{
+    private static final Logger LOG = Logging.get(ModuleAnalysisHelper.class);
+    public static final String LAST_RESORT_DEFAULT_GROUP_ID = "com.mycompany.mavenized";
+
+
+    GraphContext grCtx;
+
+
+    public ModuleAnalysisHelper(GraphContext grCtx)
+    {
+        this.grCtx = grCtx;
+    }
+
+
+    String deriveGroupId(ProjectModel projectModel)
+    {
+        if (projectModel instanceof MavenProjectModel){
+            MavenProjectModel mavenProject = (MavenProjectModel) projectModel;
+            mavenProject.getGroupId();
+        }
+
+        // User can set this via --mavenizeGroupId.
+        {
+            String groupId = (String) grCtx.getOptionMap().get(MavenizeGroupIdOption.NAME);
+            if(groupId != null)
+            {
+                if (groupId.matches(MavenizeGroupIdOption.REGEX_GROUP_ID))
+                    return groupId;
+                LOG.severe(MavenizeGroupIdOption.NAME + " doesn't match the groupId pattern, ignoring: " + groupId);
+            }
+        }
+
+        // Try to take what the user put to --packages.
+        {
+            List<String> scanPackages = (List<String>) grCtx.getOptionMap().get(ScanPackagesOption.NAME);
+            if(scanPackages != null && !scanPackages.isEmpty() && scanPackages.get(0).contains("."))
+                return scanPackages.get(0);
+        }
+
+        if (true)
+            return LAST_RESORT_DEFAULT_GROUP_ID;
+
+
+        // Get shared prefix of all packages in this application. Can result to "".
+        String fromPackages = new ModuleAnalysisHelper(grCtx).deriveGroupIdFromPackages(projectModel);
+        if (fromPackages != null)
+            return fromPackages;
+
+        return LAST_RESORT_DEFAULT_GROUP_ID;
+    }
+
+
+
+    /**
+     * Counts the packages prefixes appearing in this project
+     * and if some of them make more than half of the total of existing packages,
+     * this prefix is returned. Otherwise, returns null.
+     */
+    String deriveGroupIdFromPackages(ProjectModel projectModel)
+    {
+        Map<String,Integer> pkgsMap = new HashMap<>();
+        Set<String> pkgs = new HashSet<>(1000);
+        GremlinPipeline<Vertex, Vertex> pipeline = new GremlinPipeline<>(projectModel);
+
+        PipeFunction<String, String> upToThirdDot =
+                new PipeFunction<String, String>(){
+                    @Override public String compute(String pkgName) {
+                        int upToThirdDot = StringUtils.ordinalIndexOf(pkgName, ".", 3);
```

Comment:
This is just a helper, it isn't something really hard-setting the package. It's something to use if the user didn't specify using --mavenize.groupId, and the archive or project name is something insane, like few sencences paragraph (a description) or a number or such.